### PR TITLE
feat: wallet integration

### PR DESCRIPTION
**Pull request by peterwht**
Enables Pop CLI to allow users to sign with their browser extension wallet.

This ie enabled for the following features:
- pop up contract
- pop call contract
- pop call chain

To enable, the `--use-wallet` must be provided.

The frontend is read from a string and served for use. Pop CLI will automatically launch your browser for signing.

The frontend can be found here: https://github.com/r0gue-io/pop-wallet-signing-portal

Currently, the built frontend index.html must be pasted into pop-cli manually.

## Testing
This example has the following context:
- No running node (uses auto launched substrate-contracts-node)
- A flipper contract one directory up
```
# deploy contract (don't terminate node)
cargo run up contract --use-wallet --path=../flipper --args true
# call contract
cargo run call contract --path ../flipper --contract <CONTRACT_ADDRESS> --message flip --url ws://localhost:9944/ --use-wallet --execute

# call parachain
cargo run call parachain --use-wallet
```

## New Dependencies
Three new dependencies were added for the server:

- [axum](https://crates.io/crates/axum), built by the tokio-rs organization and is used for creating the web-API. Uses tokio. Relatively lightweight.
- [tower_http](https://crates.io/crates/tower_http), recommended for use with axum. Tower serves the static frontend.
- [open](https://crates.io/crates/open), used for auto launching the browser. Is the most downloaded and popular crate for this functionality.

### CODE REVIEW

File: crates/pop-cli/src/common/wallet.rs
Comment (AlexD10S) [inline crates/pop-cli/src/common/wallet.rs:38]: This test can be removed already

File: crates/pop-contracts/Cargo.toml
Comment (AlexD10S) [inline crates/pop-contracts/Cargo.toml:45]: This appears to be a mistake, please remove it.

**Review by AlexD10S — COMMENTED** [review]
We need to merge: https://github.com/r0gue-io/pop-cli/pull/374 into this branch, to fix the tests.

It would also be ideal to use the latest frontend (https://github.com/r0gue-io/wallet-signing-portal/pull/3) already here.

File: crates/pop-cli/src/commands/up/contract.rs
Comment (AlexD10S) [inline crates/pop-cli/src/commands/up/contract.rs:383]: 1 line comment would be great.

**Comment by evilrobot-01** [comment]
Can we get something in the PR description which shows how to use/test the feature please?

File: crates/pop-cli/src/common/wallet.rs
Comment (evilrobot-01) [inline crates/pop-cli/src/common/wallet.rs:14]: Can we change this to a spinner please? It should provide better visual feedback, but then also disappear once the signed payload is received, cleaning up the terminal a little. Otherwise there are various elements which appear as noise once done.

File: crates/pop-cli/src/commands/call/chain.rs
Comment (evilrobot-01) [inline crates/pop-cli/src/commands/call/chain.rs:259]: Seems to be very similar code as above - can we refactor into a function?

File: crates/pop-cli/src/commands/call/chain.rs
Comment (evilrobot-01) [inline crates/pop-cli/src/commands/call/chain.rs:277]: Clippy warning

File: crates/pop-cli/src/commands/call/chain.rs
Comment (evilrobot-01) [inline crates/pop-cli/src/commands/call/chain.rs:732]:
```suggestion
		.expect_confirm("Do you want to use your browser wallet to sign the extrinsic? (Selecting 'No' will prompt you to manually enter the secret key URI for signing, e.g., '//Alice')", true);
```

File: crates/pop-cli/src/commands/call/chain.rs
Comment (evilrobot-01) [inline crates/pop-cli/src/commands/call/chain.rs:875]:
```suggestion
			.expect_confirm("Do you want to use your browser wallet to sign the extrinsic? (Selecting 'No' will prompt you to manually enter the secret key URI for signing, e.g., '//Alice')", false)
```

File: crates/pop-cli/src/commands/call/contract.rs
Comment (evilrobot-01) [inline crates/pop-cli/src/commands/call/contract.rs:319]: Assume this is the same block as found in the call chain implementation. We really should have a single function so we have a consistent experience. I suspect this may be due to feature gating, in which case the shared function can be located within the call module and feature gate to either contracts or parachains.

File: crates/pop-cli/src/commands/call/contract.rs
Comment (evilrobot-01) [inline crates/pop-cli/src/commands/call/contract.rs:328]: Just checking we want the default input to be Alice? Fine for the placeholder.

File: crates/pop-cli/src/commands/call/contract.rs
Comment (evilrobot-01) [inline crates/pop-cli/src/commands/call/contract.rs:491]: This is never cancelled, leaving an unnecessary artifact on screen.

File: crates/pop-cli/src/commands/call/contract.rs
Comment (evilrobot-01) [inline crates/pop-cli/src/commands/call/contract.rs:513]:
```suggestion
			Weight::zero()
```

**Review by evilrobot-01 — COMMENTED** [review]
Partial review.

File: crates/pop-cli/src/commands/call/contract.rs
Comment (peterwht) [reply crates/pop-cli/src/commands/call/contract.rs:328]: cc @AlexD10S, seems like this is currently required for the flow of call contract?

File: crates/pop-contracts/Cargo.toml
Comment (peterwht) [reply crates/pop-contracts/Cargo.toml]: Thanks for catching this. It was used in a unit test and made its way in again. Resolved here: https://github.com/r0gue-io/pop-cli/pull/371/commits/89df5f6fedd9fe606352d19b4c156ff3aa8b3349

File: Cargo.toml
Comment (peterwht) [reply Cargo.toml:56]: in progress! Should have a cargo-contract release tomorrow.

File: crates/pop-cli/Cargo.toml
Comment (peterwht) [reply crates/pop-cli/Cargo.toml]: https://github.com/r0gue-io/pop-cli/pull/371/commits/ddb8ba6db338bfe3e1c6d091c726e65afdea9f09

File: crates/pop-cli/src/common/wallet.rs
Comment (peterwht) [reply crates/pop-cli/src/common/wallet.rs]: https://github.com/r0gue-io/pop-cli/pull/371/commits/ddb8ba6db338bfe3e1c6d091c726e65afdea9f09#diff-f1b7ff543f883f3dc2a2c8b3fc1c93e144e63e718655d11893011bc254f32e47R13

Comment (peterwht) [reply crates/pop-cli/src/common/wallet.rs]: went with request_signature

File: crates/pop-cli/src/commands/call/contract.rs
Comment (peterwht) [reply crates/pop-cli/src/commands/call/contract.rs]: https://github.com/r0gue-io/pop-cli/pull/371/commits/ddb8ba6db338bfe3e1c6d091c726e65afdea9f09#diff-60233806b7c6493fe1742695952147d4dc1bcefcbdb4886da81e63e788941bddR59

File: crates/pop-cli/src/commands/call/chain.rs
Comment (peterwht) [reply crates/pop-cli/src/commands/call/chain.rs]: Refactored here into `wallet::prompt_to_use_wallet`: https://github.com/r0gue-io/pop-cli/pull/371/commits/6035fd86b7ca679127b5cd34765a51dca6a66f13#diff-f1b7ff543f883f3dc2a2c8b3fc1c93e144e63e718655d11893011bc254f32e47R53

File: crates/pop-cli/src/commands/call/chain.rs
Comment (peterwht) [reply crates/pop-cli/src/commands/call/chain.rs]: https://github.com/r0gue-io/pop-cli/pull/371/commits/ddb8ba6db338bfe3e1c6d091c726e65afdea9f09#diff-00e3547545bcb69419c2e4ce2de7d2e7387b839e1020fbbf6618928b8ee61478R277

File: crates/pop-cli/src/commands/call/chain.rs
Comment (peterwht) [reply crates/pop-cli/src/commands/call/chain.rs]: Refactored here into wallet::prompt_to_use_wallet: https://github.com/r0gue-io/pop-cli/commit/6035fd86b7ca679127b5cd34765a51dca6a66f13#diff-f1b7ff543f883f3dc2a2c8b3fc1c93e144e63e718655d11893011bc254f32e47R53

File: crates/pop-cli/src/commands/call/chain.rs
Comment (peterwht) [reply crates/pop-cli/src/commands/call/chain.rs]: Refactored here into wallet::prompt_to_use_wallet: https://github.com/r0gue-io/pop-cli/commit/6035fd86b7ca679127b5cd34765a51dca6a66f13#diff-f1b7ff543f883f3dc2a2c8b3fc1c93e144e63e718655d11893011bc254f32e47R53.

Note: there is a follow-up commit changing to extrinsic instead of transaction.

File: crates/pop-cli/src/commands/call/contract.rs
Comment (peterwht) [reply crates/pop-cli/src/commands/call/contract.rs]: https://github.com/r0gue-io/pop-cli/pull/371/commits/6035fd86b7ca679127b5cd34765a51dca6a66f13

File: crates/pop-cli/src/commands/call/contract.rs
Comment (peterwht) [reply crates/pop-cli/src/commands/call/contract.rs:493]: https://github.com/r0gue-io/pop-cli/pull/371/commits/ddb8ba6db338bfe3e1c6d091c726e65afdea9f09#diff-60233806b7c6493fe1742695952147d4dc1bcefcbdb4886da81e63e788941bddR420

File: crates/pop-cli/src/commands/call/contract.rs
Comment (peterwht) [reply crates/pop-cli/src/commands/call/contract.rs]: https://github.com/r0gue-io/pop-cli/pull/371/commits/ddb8ba6db338bfe3e1c6d091c726e65afdea9f09#diff-60233806b7c6493fe1742695952147d4dc1bcefcbdb4886da81e63e788941bddR514

File: crates/pop-cli/src/commands/call/chain.rs
Comment (peterwht) [reply crates/pop-cli/src/commands/call/chain.rs]: refactored into `determine_signing_method`: https://github.com/r0gue-io/pop-cli/pull/371/commits/6035fd86b7ca679127b5cd34765a51dca6a66f13

File: crates/pop-cli/src/common/wallet.rs
Comment (peterwht) [reply crates/pop-cli/src/common/wallet.rs]: https://github.com/r0gue-io/pop-cli/pull/371/commits/6035fd86b7ca679127b5cd34765a51dca6a66f13#diff-f1b7ff543f883f3dc2a2c8b3fc1c93e144e63e718655d11893011bc254f32e47R28

File: crates/pop-cli/src/commands/call/chain.rs
Comment (peterwht) [reply crates/pop-cli/src/commands/call/chain.rs]: https://github.com/r0gue-io/pop-cli/pull/371/commits/6035fd86b7ca679127b5cd34765a51dca6a66f13#diff-f1b7ff543f883f3dc2a2c8b3fc1c93e144e63e718655d11893011bc254f32e47R28

File: crates/pop-cli/src/commands/call/contract.rs
Comment (AlexD10S) [reply crates/pop-cli/src/commands/call/contract.rs]: Yes, I think it’s useful to have a default value here. In my experience, most of the time I’ve used `//Alice`.

File: crates/pop-cli/src/wallet_integration.rs
Comment (al3mart) [inline crates/pop-cli/src/wallet_integration.rs:73]: Let's have a port as one of the parameters! We can that way always make sure we are using an available port.

File: crates/pop-cli/src/common/wallet.rs
Comment (al3mart) [inline crates/pop-cli/src/common/wallet.rs:24]: Maybe using `new_with_address` with an available port or refactoring `new` to be able to pass a port number would be necessary before merging.

**Review by al3mart — CHANGES_REQUESTED** [review]

File: Cargo.toml
Comment (AlexD10S) [reply Cargo.toml:56]: Done, @peterwht you can use "5.0.2" in the 3 crates

Comment (peterwht) [reply Cargo.toml:56]: Alright!

**Comment by peterwht** [comment]
to comment here, from @evilrobot-01, would be nice to auto launch the browser.

File: crates/pop-cli/src/wallet_integration.rs
Comment (peterwht) [reply crates/pop-cli/src/wallet_integration.rs]: https://github.com/r0gue-io/pop-cli/pull/371/commits/0e19b8a6cc693a545ffa9cd62eaff9e744127fb6

File: crates/pop-cli/src/common/wallet.rs
Comment (peterwht) [reply crates/pop-cli/src/common/wallet.rs]: https://github.com/r0gue-io/pop-cli/pull/371/commits/0e19b8a6cc693a545ffa9cd62eaff9e744127fb6

**Review by al3mart — COMMENTED** [review]
Have reviewed the last commits since I last left some feedback here.
Everything is looking pretty solid to me.

We most definitely need to address the sleep times set for CI in tests in the future and see if we can find some work around them.

Happy to proceed with approval if we aren't planning on pushing more changes here. I'll hold for now just in case.

File: crates/pop-cli/tests/contract.rs
Comment (AlexD10S) [inline crates/pop-cli/tests/contract.rs:36]: As a future refactor, we make this struct available from `pop-cli` and remove the need of duplication.

**Review by AlexD10S — APPROVED** [review]
Great job on this once again! 
Aside from a few tweaks that can be done to improve the frontend, everything here looks great to me.

**Comment by evilrobot-01** [comment]
FYI: https://github.com/r0gue-io/pop-cli/pull/390

I also think we should simplify the PR title to "feat: wallet integration" before squash merging.

**Review by evilrobot-01 — COMMENTED** [review]
See https://github.com/r0gue-io/pop-cli/pull/390 for review in the form of suggested improvments to get this finalised. Happy to approve this once reviewed/addressed/incorporated.

File: crates/pop-cli/src/common/contracts.rs
Comment (AlexD10S) [reply crates/pop-cli/src/common/contracts.rs:75]: The coverage for console outputs of  `pop up contract`, including this method, will be fully addressed once this PR https://github.com/r0gue-io/pop-cli/pull/315 is complete and merged.
In this PR, we only moved it to a common file.

File: crates/pop-cli/src/common/wallet.rs
Comment (peterwht) [reply crates/pop-cli/src/common/wallet.rs:24]: If 9090 is taken, it will choose a random port. The UX is almost exactly the same. The only difference is that the last wallet used isn't saved between different ports (different URLs)

**Review by AlexD10S — APPROVED** [review]

**Review by evilrobot-01 — APPROVED** [review]
Amazing, so glad to finally see this completed. Well done to all involved! 🙌

**Review by al3mart — APPROVED** [review]
Let's get this merged :fire:

**Review by AlexD10S — APPROVED** [review]

**Pull request merged**
